blockly-rtc/src/WorkspaceClient.js

```diff
@@ -155,60 +153,69 @@ export default class WorkspaceClient {
       if (rows.length == 0) {
         return eventQueue;
       };
-  
+
       this.lastSync = rows[rows.length - 1].serverId;
   
       // No local changes.
       if (this.notSent.length == 0 && this.inProgress.length == 0) {
         rows.forEach((row) => {
-          eventQueue.push(this.createWorkspaceAction_(row.event, true));
+          eventQueue.push.apply(
+            eventQueue, this.createWorkspaceActions_(row.events, true));
         });
         return eventQueue;
       };
-    
+  
       // Common root, remove common events from server events.
       if (this.inProgress.length > 0 && rows[0].entryId == this.inProgress[0].entryId) {
-        rows = rows.slice(this.inProgress.length);
+        rows.shift();
         this.inProgress = [];
       };
   
       if (rows.length > 0) {
         // Undo local events.
-        this.notSent.slice().reverse().forEach((row) => {
-          eventQueue.push(this.createWorkspaceAction_(row.event, false));
-        });
-        this.inProgress.slice().reverse().forEach((row) => {
-          eventQueue.push(this.createWorkspaceAction_(row.event, false));
-        });
+        eventQueue.push.apply(
+          eventQueue,
+          this.createWorkspaceActions_(this.notSent.slice().reverse(), false));
+        if (this.inProgress.length > 0) {
+          eventQueue.push.apply(eventQueue,this.createWorkspaceActions_(
+            this.inProgress[0].events.slice().reverse(), false));
+        };
         // Apply server events.
         rows.forEach((row) => {
-          eventQueue.push(this.createWorkspaceAction_(row.event, true));
+          eventQueue.push.apply(
+            eventQueue, this.createWorkspaceActions_(row.events, true));
           if (this.inProgress.length > 0 && row.entryId == this.inProgress[0].entryId) {
             this.inProgress.shift();
           };
         });
         // Reapply remaining local changes.
-        this.inProgress.forEach((row) => {
-          eventQueue.push(this.createWorkspaceAction_(row.event, true));
-        });
-        this.notSent.forEach((row) => {
-          eventQueue.push(this.createWorkspaceAction_(row.event, true));
-        });
+        if (this.inProgress.length > 0) {
+          eventQueue.push.apply(
+            eventQueue,
+            this.createWorkspaceActions_(this.inProgress[0].events, true));
```

Comment:
Is the queue guaranteed to be of length 1 here?

Reply:
Yes, at this place in the code the eventQueue is guaranteed to be of at least length 1.
The only case I can think of for it not being at least 1 is if the all the events it got from the server were created by the client, in which case the workspace is up to date.

Or were you talking about `this.inProgress` being of length 1? In which case yes. At this point `this.inProgress` can only be of length 0 or 1. If it is of length 0, the function would have exit out at [L165](https://github.com/google/blockly-samples/pull/21/files/a799c08f6dd2860a3fc67b85e212152fe5780e92#diff-039d24c9cc7f1f76cd5c17b18c7cdbb7L165)

Reply:
I meant `this.inProgress`.  I see where it'll exit if it's of length 0.  If the length is only ever 0 or 1, why store it as an array instead of a nullable object?

Reply:
Discussed offline.